src/library/basesqltablemodel.cpp

```diff
@@ -419,19 +375,102 @@ void BaseSqlTableModel::setSort(int column, Qt::SortOrder order) {
+            }
+        }
+
+        // set new sort as head and shift out old sort
+        SortColumn sc;
+        sc.column = column;
+        sc.order = order;
+        m_sortColumns.prepend(sc);
+
+        if (m_sortColumns.size() > 3) {
+            m_sortColumns.removeLast();
+        }
+    }
 
-    m_iSortColumn = column;
-    m_eSortOrder = order;
+    // we have two selects for sorting, since keeping the select history
+    // across the two selects is hard, we do this only for the trackSource
+    // this is OK, because the coloums of the table are virtual in cas of
+    // preview column or individual like playlist track number so that we
+    // do not need the history anyway.
+
+    // reset the old order by clauses
+    m_trackSourceOrderBy.clear();
+    m_tableOrderBy.clear();
+    m_trackSourceSortColumn = 0;
+    m_trackSourceSortOrder = Qt::AscendingOrder;
+
+    if (column > 0 && column < m_tableColumns.size()) {
+        // Table sorting, no history
+        m_tableOrderBy.append(QLatin1String("ORDER BY "));
+        QString field = m_tableColumns[column];
+        QString sort_field = QString("%1.%2").arg(m_tableName, field);
+        m_tableOrderBy.append(sort_field);
+    #ifdef __SQLITE3__
+        m_tableOrderBy.append(QLatin1String(" COLLATE localeAwareCompare"));
+    #endif
+        m_tableOrderBy.append((order == Qt::AscendingOrder) ?
+                QLatin1String(" ASC") :
+                QLatin1String(" DESC"));
+        m_sortColumns.clear();
+    } else if (m_trackSource) {
+        foreach (SortColumn sc, m_sortColumns) {
+            // TrackSource Sorting, current sort + two from history
+            if (!m_trackSourceOrderBy.isEmpty()) {
+                // second cycle
+                m_trackSourceOrderBy.append(QLatin1String(", "));
+            }
+            QString field;
+            if (sc.column == 0) {
+                field = m_trackSource->columnNameForFieldIndex(0);
+            } else {
+                // + 1 to skip id coloumn
+                int ccColumn = sc.column - m_tableColumns.size() + 1;
+                field = m_trackSource->columnNameForFieldIndex(ccColumn);
+                if (m_trackSourceOrderBy.isEmpty()) {
+                    // first cycle
+                    m_trackSourceSortColumn = ccColumn;
+                    m_trackSourceSortOrder = sc.order;
+                }
+            }
+            QString sort_field = field;
+            m_trackSourceOrderBy.append(sort_field);
+
+    #ifdef __SQLITE3__
+            m_trackSourceOrderBy.append(QLatin1String(" COLLATE localeAwareCompare"));
+    #endif
+            m_trackSourceOrderBy.append((sc.order == Qt::AscendingOrder) ?
+                    QLatin1String(" ASC") :
+                    QLatin1String(" DESC"));
+            qDebug() << m_trackSourceOrderBy << sc.order;
+        }
+        if (!m_trackSourceOrderBy.isEmpty()) {
+            m_trackSourceOrderBy.prepend(QLatin1String("ORDER BY "));
```

Comment:
I think all of these strings can just be plain literals, we have plenty of places in the code where that call does not appear and the db works fine.